reddit.py

```diff
@@ -0,0 +1,141 @@
+"""Reddit source code and datastore model classes.
+
+"""
+import logging
+
+from granary import reddit as gr_reddit
+from granary import source as gr_source
+from oauth_dropins import reddit as oauth_reddit
+from oauth_dropins.webutil.util import json_dumps, json_loads
+import webapp2
+from webob import exc
+
+import models
+import util
+
+
+class Reddit(models.Source):
+  """A Reddit account.
+
+  The key name is the username.
+  """
+  GR_CLASS = gr_reddit.Reddit
+  OAUTH_START_HANDLER = oauth_reddit.StartHandler
+  SHORT_NAME = 'reddit'
+  TYPE_LABELS = {
+    'post': 'submission',
+    'comment': 'comment',
+  }
+  TRANSIENT_ERROR_HTTP_CODES = ('404',)
+
+  @staticmethod
+  def new(handler, auth_entity=None, **kwargs):
+    """Creates and returns a :class:`Twitter` entity.
+
+    Args:
+      handler: the current :class:`webapp2.RequestHandler`
+      auth_entity: :class:`oauth_dropins.twitter.TwitterAuth`
+      kwargs: property values
+    """
+    user = json_loads(auth_entity.user_json)
+    gr_source = gr_reddit.Reddit(auth_entity.refresh_token)
+    return Reddit(id=user.get('name'),
+                  auth_entity=auth_entity.key,
+                  url=gr_source.user_url(user.get('name')),
+                  name=user.get('name'),
+                  picture=user.get('icon_img'),
+                  **kwargs)
+
+  def silo_url(self):
+    """Returns the Twitter account URL, e.g. https://twitter.com/foo."""
+    return self.gr_source.user_url(self.key_id())
+
+  def label_name(self):
+    """Returns the username."""
+    return self.key_id()
+
+  def search_for_links(self):
+    """Searches for activities with links to any of this source's web sites.
+
+    Twitter search supports OR:
+    https://dev.twitter.com/rest/public/search
+
+    ...but it only returns complete(ish) results if we strip scheme from URLs,
+    ie search for example.com instead of http://example.com/, and that also
+    returns false positivies, so we check that the returned tweets actually have
+    matching links. https://github.com/snarfed/bridgy/issues/565
+
+    Returns:
+      sequence of ActivityStreams activity dicts
+    """
+    urls = set(util.schemeless(util.fragmentless(url), slashes=False)
```

Comment:
does reddit search also need the scheme removed from URLs? otherwise try the full URLs, like normal?

Reply:
i'm honestly not sure what scheme means here. but if you mean the http part I think it needs to be removed.

Reply:
yeah scheme is protocol. you can probably test searches for different URL formats in reddit's UI if you want.

Reply:
with http in the search, it doesn't find it, so I'll use the schemeless.